ci-operator/step-registry/ipi/conf/azure/osimage/ipi-conf-azure-osimage-commands.sh

```diff
@@ -58,56 +58,64 @@ AZURE_AUTH_TENANT_ID="$(<"${AZURE_AUTH_LOCATION}" jq -r .tenantId)"
 
 # log in with az
 az login --service-principal -u "${AZURE_AUTH_CLIENT_ID}" -p "${AZURE_AUTH_CLIENT_SECRET}" --tenant "${AZURE_AUTH_TENANT_ID}" --output none
-# random select the market-place image
-offer="rh-ocp-worker"
-images=$(az vm image list --all --offer ${offer} --publisher redhat -o tsv --query "[?starts_with(urn,'RedHat:${offer}:${offer}')].urn")
-mapfile -t marketImages <<< "${images}"
-echo "choice market-place image from:" "${marketImages[@]}"
-imageTotal=${#marketImages[@]}
-echo "Image total: ${imageTotal} "
-if [ ${imageTotal} -lt 1 ] ; then
-  echo "Fail to find the market-place image under the ${AZURE_AUTH_LOCATION}"
-  exit 1
+
+#Get imageinfo list
+if [[ "${OS_IMAGE_URN}" != "" ]]; then
+    image=${OS_IMAGE_URN}
+    IFS=':' read -ra imageInfo <<< "${image}"
+else
+    # random select image from azure marketplace with sepcified version
+    # Generation: gen1 or gen2
+    offer="rh-ocp-worker"
+    publisher="RedHat"
+    images=$(az vm image list --all --offer ${offer} -o tsv --query "[?publisher=='${publisher}'] | [?version=='${OS_IMAGE_VERSION}'].urn")
+    mapfile -t marketImages <<< "${images}"
+    echo "choice market-place image from:" "${marketImages[@]}"
+    imageTotal=${#marketImages[@]}
+    echo "Image total: ${imageTotal} "
+    if [ ${imageTotal} -lt 1 ] ; then
+        echo "Fail to find the market-place image under the ${AZURE_AUTH_LOCATION}"
+        exit 1
+    fi
+    selected_image_idx=$((RANDOM % ${imageTotal}))
+    image=${marketImages[selected_image_idx]}
+    echo "$(date -u --rfc-3339=seconds) - Selected image ${image}"
+    IFS=':' read -ra imageInfo <<< "${image}"
 fi
-selected_image_idx=$((RANDOM % ${imageTotal}))
-image=${marketImages[selected_image_idx]}
-echo "$(date -u --rfc-3339=seconds) - Selected image ${image}"
-IFS=':' read -ra imageInfo <<< "${image}"
 echo "imageInfo: " "${imageInfo[@]}"
 
-# create a patch to set osImage for compute
-PATCH="/tmp/install-config-existingworkers-marketimage.yaml.patch"
+# image plan is case-sensitive, make sure that publisher/offer/sku keep the same as plan
+if [[ "${OS_IMAGE_PLAN}" == "WithPurchasePlan" ]] || [[ "${OS_IMAGE_PLAN}" == "" ]]; then
+    imageInfo[0]=$(az vm image show --urn ${image} --query 'plan.publisher' -otsv)
+    imageInfo[1]=$(az vm image show --urn ${image} --query 'plan.product' -otsv)
+    imageInfo[2]=$(az vm image show --urn ${image} --query 'plan.name' -otsv)
+fi
+
+#set plan for os image, default value is WithPurchasePlan
+if [[ "${OS_IMAGE_PLAN}" == "NoPurchasePlan" ]]; then
+    os_image_plan="plan: ${OS_IMAGE_PLAN}"
+fi
+
+# create/apply a patch to set osImage for compute
+PATCH="/tmp/install-config-workers-marketimage.yaml.patch"
 cat > "${PATCH}" << EOF
 compute:
 - platform:
     azure:
       osImage:
-        publisher: redhat
+        publisher: imageInfo[0]
         offer: ${imageInfo[1]}
         sku: ${imageInfo[2]}
         version: ${imageInfo[3]}
+        ${os_image_plan}
```

Comment:
We need to set `os_image_plan` with an empty value in previous lines when `OS_IMAGE_PLAN` is empty?

And when `OS_IMAGE_PLAN` is set to `WithPurchasePlan`, we also assemble `os_image_plan`, right?

Reply:
yes, fixed.